### PR TITLE
dev-build/kas: new package, add 4.5

### PR DESCRIPTION
**Pull request by martinhomuth**
kas is a setup tool for bitbake based projects.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by thesamesam** [comment]
I just pushed a fix for the CI failure (unrelated to your PR), please rebase.

File: app-misc/kas/kas-4.5.ebuild [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_USE_PEP517=setuptools
+PYTHON_COMPAT=( python3_{6..13} )
```
Comment (thesamesam) [inline app-misc/kas/kas-4.5.ebuild:7]: < 3.10 is a noop in Gentoo:
```suggestion
PYTHON_COMPAT=( python3_{10..13} )
```

File: app-misc/kas/kas-4.5.ebuild [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+		-> ${MY_P}.gh.tar.gz
+"
+S=${WORKDIR}/${MY_P}
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64"
+
+RDEPEND="
+	>=dev-python/pyyaml-3.0.0
```
Comment (thesamesam) [inline app-misc/kas/kas-4.5.ebuild:25]: These deps aren't right, see:
* https://projects.gentoo.org/python/guide/distutils.html#dependencies
* https://projects.gentoo.org/python/guide/multi.html#dependencies

File: app-misc/kas/kas-4.5.ebuild [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+EAPI=8
+
+DISTUTILS_USE_PEP517=setuptools
+PYTHON_COMPAT=( python3_{6..13} )
+
+inherit distutils-r1
+
+MY_P=kas-${PV}
+DESCRIPTION="Setup tool for bitbake based projects."
```
Comment (thesamesam) [inline app-misc/kas/kas-4.5.ebuild:12]:
```suggestion
DESCRIPTION="Setup tool for bitbake based projects"
```

File: app-misc/kas/metadata.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+<pkgmetadata>
+  <maintainer type="person">
+    <email>mh@emlix.com</email>
+    <name>Martin Homuth</name>
+  </maintainer>
+  <maintainer type="project" proxied="proxy">
+    <email>proxy-maint@gentoo.org</email>
+    <name>Proxy Maintainers</name>
+  </maintainer>
+  <upstream>
```
Comment (thesamesam) [inline app-misc/kas/metadata.xml:12]: Can you add the pypi remote-id too please, if it's available there?

**Review by thesamesam — CHANGES_REQUESTED** [review]

File: app-misc/kas/metadata.xml (hunk 4 above)
Comment (martinhomuth) [reply app-misc/kas/metadata.xml:12]: Thanks for the hint, switched to pypi.eclass while at it.

File: app-misc/kas/kas-4.5.ebuild (hunk 2 above)
Comment (martinhomuth) [reply app-misc/kas/kas-4.5.ebuild]: I think I use it correctly now

**Comment by thesamesam** [comment]
Looking good - many thanks!

**Comment by thesamesam** [comment]
A final quick question -- do we want to do `dev-embedded/` instead?

**Comment by martinhomuth** [comment]
> A final quick question -- do we want to do `dev-embedded/` instead?

As yocto does not strictly focuses on embedded systems to be created I would like to decline. However, while I am pretty new to this, categorizing this as with more development nature would be good. So I am not sure..

**Comment by thesamesam** [comment]
It looks like test fail with pypi as the tests aren't in the tarball.

**Comment by martinhomuth** [comment]
> It looks like test fail with pypi as the tests aren't in the tarball.

Shoot, should have tested that. If that is the case would you prefer switching to the github source or dropping the tests?

**Comment by thesamesam** [comment]
Let's switch to GH if that's OK, thanks

**Comment by thesamesam** [comment]
Gah, it looks like there's a missing test dep on `dev-tcltk/snack`:
```
--- a/app-misc/kas/kas-4.5.ebuild
+++ b/app-misc/kas/kas-4.5.ebuild
@@ -4,7 +4,7 @@
 EAPI=8

 DISTUTILS_USE_PEP517=setuptools
-PYTHON_COMPAT=( python3_{10..13} )
+PYTHON_COMPAT=( python3_{10..12} )

 inherit distutils-r1

@@ -35,7 +35,10 @@ RDEPEND="

 BDEPEND="
        ${RDEPEND}
-       test? ( dev-vcs/mercurial )
+       test? (
+               dev-vcs/mercurial
+               dev-tcltk/snack[python,${PYTHON_USEDEP}]
+       )
 "

 distutils_enable_tests pytest
```

but snack doesn't build because of some modern C bits which needs investigation...

File: app-misc/kas/kas-4.5.ebuild [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+RDEPEND="
+	>=dev-python/pyyaml-3.0.0[${PYTHON_USEDEP}]
+	>=dev-python/distro-1.0.0[${PYTHON_USEDEP}]
+	>=dev-python/jsonschema-2.5.0[${PYTHON_USEDEP}]
+	>=dev-python/kconfiglib-14.1.0[${PYTHON_USEDEP}]
+	>=dev-python/GitPython-3.1.0[${PYTHON_USEDEP}]
+"
+
+PROPERTIES="test_network"
```
Comment (thesamesam) [inline app-misc/kas/kas-4.5.ebuild:36]: Can you add a comment here summarising your efforts? (e.g. "Tests can work offline but need some wrangling to accept offline clones of git repo. TODO") 

(I would just add it myself but I'm going to figure out what's going on with snack, so it's not like I'm going to merge it right this second anyway.)

**Comment by thesamesam** [comment]
Thanks for looking. I'll figure out snack.

**Comment by thesamesam** [comment]
One more comment, sorry!

snack is used not just in tests (duh), but in a plugin: [kas/plugins/menu.py](https://github.com/siemens/kas/blob/247d9c24be93bae57b43a57ef1e5675cfc89914a/kas/plugins/menu.py#L85). It's up to you if you want to be an optfeature, unconditional dependency, or have a USE flag. We discourage USE flags for runtime-only deps (where it doesn't change the built package) but for pure-Python stuff, we admit a bit of leeway.

**Comment by martinhomuth** [comment]
Moved kas to `dev-build` where it belongs

**Comment by thesamesam** [comment]
Great job!

**Comment by henning-schild** [comment]
Interesting to see that this package made it into gentoo. I happen to know a thing or two about it. While it can be used as is users should rather be pointed to https://github.com/siemens/kas/blob/master/kas-container. I used to have that script as a gentoo package which would depend on docker or podman.

https://kas.readthedocs.io/en/latest/userguide/kas-container.html the decoupling from the host is key to make sure a build that succeeds on one machine will work the same way on another, a gentoo ebuild for kas partially defeats that decoupling. So if i was wearing a kas-hat i would suggest using the debian-based containers always and only package the container script on other distros.

Anyhow an ebuild called kas-container just holding the most recent version of the script would be a nice addition, IMHO it should be the only ebuild around kas.

File: dev-build/kas/kas-4.5.ebuild [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	>=dev-python/jsonschema-2.5.0[${PYTHON_USEDEP}]
+	>=dev-python/kconfiglib-14.1.0[${PYTHON_USEDEP}]
+	>=dev-python/GitPython-3.1.0[${PYTHON_USEDEP}]
+	dev-tcltk/snack[python,${PYTHON_USEDEP}]
+"
+
+BDEPEND="
+	${RDEPEND}
+	test? (
+			dev-vcs/mercurial
```
Comment (henning-schild) [inline dev-build/kas/kas-4.5.ebuild:44]: In fact git and mercurial should be optfeatures, kas is - among other things - mainly about cloning repos.

**Comment by thesamesam** [comment]
@henning-schild I think your tone could be moderated a bit. Anyway, thanks for looking.

File: dev-build/kas/kas-4.5.ebuild [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+DISTUTILS_USE_PEP517=setuptools
+PYTHON_COMPAT=( python3_{10..12} )
+
+inherit distutils-r1
+
+MY_P="kas-${PV}"
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-build/kas/kas-4.5.ebuild:11]: It changed when pypi.eclass was originally being used. I'll clean it up.

File: dev-build/kas/kas-4.5.ebuild (hunk 6 above)
Comment (thesamesam) [reply dev-build/kas/kas-4.5.ebuild:44]: I don't think that's an "in fact" -- really, many test deps often double up as test deps _as well_. I'll add that in though.

**Pull request closed without merging**